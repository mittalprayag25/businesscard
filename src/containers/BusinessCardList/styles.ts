import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  list: {
    marginTop: 20,
    width: '90%',
  },
  menuItem: {height: 20, width: 20, marginRight: 10},
});
export default styles;
