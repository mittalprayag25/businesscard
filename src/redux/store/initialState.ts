const initialState = {
  businessCards: {
    savedCards: [],
  },
};

export default initialState;
