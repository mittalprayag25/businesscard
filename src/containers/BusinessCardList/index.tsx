import React, { useEffect, memo } from 'react';
import { SafeAreaView, Image, View, TouchableOpacity } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { UserDetail } from '../BusinessCardDetails';
import EmptyDetails from '../../components/EmptyDetails';
import SavedDetails from '../../components/SavedDetails';
import { useNavigation } from '@react-navigation/native';
import actions from '../../redux/actions';
import { ScrollView } from 'react-native-gesture-handler';
import { RootState } from '../../redux/reducers';
import { Routes } from './../../constants/NavigationUtils';
import styles from './styles';

const BusinessCardList = (): JSX.Element => {
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const { savedCards }: any = useSelector((s: RootState) => s.businessCards);

  const onPresshandler = () => {
    navigation.navigate(Routes.BUSINESSCARDDETAILS, {});
  };

  const onDelete = (index: number) => {
    dispatch(actions.businessList.delete(index));
  };

  const openDetails = data => {
    navigation.navigate(Routes.BUSINESSCARDDETAILS, { card: data });
  };

  useEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <TouchableOpacity onPress={onPresshandler} testID="addUser">
          <Image
            source={require('../../assets/add-user.png')}
            style={styles.menuItem}
          />
        </TouchableOpacity>
      ),
    });
  }, [navigation]);

  return (
    <>
      <SafeAreaView style={styles.container}>
        {savedCards === undefined || savedCards.length === 0 ? (
          <View style={styles.container}>
            <EmptyDetails isListEmpty={savedCards.length === 0} />
          </View>
        ) : (
          <ScrollView style={styles.list}>
            {savedCards &&
              savedCards.map(
                (savedCard: UserDetail, index: number) =>
                  savedCard && (
                    <SavedDetails
                      key={index}
                      index={index}
                      onDeleteHandler={onDelete}
                      userDetails={savedCard}
                      onOpen={openDetails}
                    />
                  ),
              )}
          </ScrollView>
        )}
      </SafeAreaView>
    </>
  );
};

export default memo(BusinessCardList);
