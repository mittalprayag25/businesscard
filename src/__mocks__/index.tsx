import configureStore from 'redux-mock-store';

import initialState from '../redux/store/initialState';

export const mockStore = configureStore()(initialState);
export const createMockStore = (state = initialState) =>
  configureStore()(state);

//mock navigation
const state = { params: {} };
export const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
  isFocused: () => true,
  setParams: (params: any) => (state.params = params),
  //getParam: (key: any) => state.params[key],
  state,
  pop: jest.fn(),
  popToTop: jest.fn(),
  reset: jest.fn(),
  replace: jest.fn(),
  setOptions: jest.fn(),
};

export const mockNavigation = {
  isFocused: () => true,
  navigate: jest.fn(),
  goBack: jest.fn(),
  pop: jest.fn(),
  getParam: jest.fn(),
  popToTop: jest.fn(),
  setOptions: jest.fn(),
  setParams: jest.fn(),
  replace: jest.fn(),
  addListener: jest.fn((event, callback) => {
    switch (event) {
      case 'didFocus':
        callback();
        return {
          remove: jest.fn(),
        };
      case 'didBlur':
        callback();
        return {
          remove: jest.fn(),
        };
      default:
        break;
    }
  }),
};

export const mockNavigationWithParams = (params: any) => ({
  ...mockNavigation,
  getParam: (key: string) => params[key],
  state: { params },
});
