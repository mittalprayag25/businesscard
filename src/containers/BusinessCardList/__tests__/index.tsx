import { render } from '@testing-library/react-native';
import React from 'react';
import * as reactRedux from 'react-redux';
import { Provider } from 'react-redux';

import { createMockStore } from '../../../__mocks__/index';
import BusinessCardList from '../index';
import initialState from './../../../redux/store/initialState';

const mockedDispatch = jest.fn();
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useFocusEffect: () => jest.fn(),
    useNavigation: () => ({
      navigate: jest.fn(),
      setOptions: jest.fn(),
    }),
  };
});

describe('Business Card List', () => {
  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <BusinessCardList />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});

describe('Business Card List with saved Cards', () => {
  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      savedCards: [
        {
          name: 'Test',
          occupation: 'IT',
          company: 'Affinidi',
          email: 'aa@bb.com',
          phone: '98765437',
          linkedin: '',
        },
      ],
    });
    jest.setTimeout(10000);
  });
  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <BusinessCardList />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
