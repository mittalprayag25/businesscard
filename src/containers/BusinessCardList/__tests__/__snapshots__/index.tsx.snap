// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Business Card List should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text>
      List is empty
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`Business Card List should match snapshot 2`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "marginTop": 20,
        "width": "90%",
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 10,
              "elevation": 8,
              "paddingHorizontal": 20,
              "paddingTop": 10,
              "shadowColor": "white",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 6,
            },
            Object {
              "backgroundColor": "white",
              "height": 200,
              "marginTop": 10,
              "width": "100%",
            },
          ]
        }
      >
        <RCTScrollView>
          <View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              Test
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              IT
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              Affinidi
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              aa@bb.com
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              98765437
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              
            </Text>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`Business Card List with saved Cards should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "marginTop": 20,
        "width": "90%",
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 10,
              "elevation": 8,
              "paddingHorizontal": 20,
              "paddingTop": 10,
              "shadowColor": "white",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 6,
            },
            Object {
              "backgroundColor": "white",
              "height": 200,
              "marginTop": 10,
              "width": "100%",
            },
          ]
        }
      >
        <RCTScrollView>
          <View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 20,
                      "marginHorizontal": 5,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              Test
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              IT
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              Affinidi
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              aa@bb.com
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              98765437
            </Text>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "paddingVertical": 5,
                }
              }
            >
              
            </Text>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
